compiler/generator_rust.c

```diff
@@ -939,6 +945,7 @@ static void generate_dollar(struct generator * g, struct node * p) {
               "~Menv.limit = env.current.len();~N", p);
     generate(g, p->left);
     if (!g->unreachable) {
+        g->V[0] = p->name;
         /* Update string variable. */
         w(g, "~M~V0 = env.current.clone().into_owned();~N");
```

Comment:
Then I don't understand the generated c code correctly.
Tried to annotate to my understanding:
```c
        //Copy env
        struct SN_env env2 = * z; /* $ s, line 5 */
        // Set current value to S[0] = 'foo'
        z->p = z->S[0];
        z->lb = z->c = 0;
        z->l = SIZE(z->p);
        {   int ret;
            {   int saved_c = z->c;
                // Set current value to 'bar'
                ret =insert_s(z, z->c, z->l, 3, s_1); /* =, line 5 */
                z->c = saved_c;
            }
            if (ret < 0) return ret;
        }
        {
            //Copy env
            struct SN_env env3 = * z; /* $ t, line 5 */
            //Set current value to S[1] = '1'
            z->p = z->S[1];
            z->lb = z->c = 0;
            z->l = SIZE(z->p);
            {   int ret;
                {   int saved_c = z->c;
                    // Set current value to S[0] (currently still 'foo')
                    ret =insert_v(z, z->c, z->l, z->S[0]); /* = s, line 5 */
                    z->c = saved_c;
                }
                if (ret < 0) return ret;
            }
            failure = 0; /* mark success */
            // Set S[1] = current value = 'foo'
            z->S[1] = z->p;
            // Reset env and thus current value to 'bar'
            * z = env3;
            if (failure) goto lab1;
        }
        failure = 0; /* mark success */
    lab1:
        //Set S[0] to current value 'bar'
        z->S[0] = z->p;
        //Reset env
        * z = env2;
```

What I probably get wrong is how pointers are handled.
```c
 z->p = z->S[0];
```
Makes `z->p` point to `z->S[0]`. So if `z->p` is changed `z->S[0]` is changed too, correct?
But then, the reset would be useless:
```c
z->S[0] = z->p;
```

Possible solutions:

1. One way to get the intended behaviour is to special case string variable assignments (this would involve work in the tokenizer and analyzer and probably the other backend which is probably not feasable).

2. Another way would be to hack up the rust implementation to use mutable pointers like the C implemenation (IMO we should avoid that).

3. Or we keep a field in the generator to what string variable the env is currently set.  
Then in `generate_assignfrom` not only the `env` could be assigned to but also the string variable.
(This would involve adding a new field to the generator struct)

I personally favor option 3 but am not sure if I understand the consequences of introducing a new field to the generator struct.